docs/naps/4-async-slicing.md

```diff
@@ -477,15 +525,38 @@ not work when layer data is mutated.
 There are some other rough edges and plenty of ways to break it, but basic
 functionality should work.
 
+The rough implementation plan is as follows.
+
+- Implement pure refactors related to slicing to support this work (no public API or behavior changes).
+- Implement high-level asynchronous slicing model described here (i.e. add `LayerSlicer`, changes to `ViewerModel`), but with existing layer slicing implementation (i.e. `Layer._set_view_slice`).
+- Incrementally add layer slicing implementation described here (i.e. `Layer._get_slice`) for each layer separately.
+
+This allows us to merge small pieces of work into `main`, testing and fixing
+them as we go. The alternative is to maintain a large branch/fork of napari that
+may be less tested and harder to merge in the end.
 
 ## Backward compatibility
 
 ### Breaks synchronous slicing behavior
 
 The main goal of this project is to perform slicing asynchronously, so it's
 natural that we might break anyone that was depending on slicing being synchronous.
-At a minimum, we must provide a public way to achieve the same fundamental goals,
-such as connecting to the `slice_ready` event.
+In particular, users or plugins that rely on `Viewer.screenshot` to capture the
+contents of the canvas after updating things like `Dims.current_step` will be affected.
+
+Our proposal will break that usage in general, but we plan to offer some public way to
+force synchronous slicing. For example, one promising idea is to define a context manager
+that would temporarily force `LayerSlicer` to always wait for asynchronous tasks to finish
+and could be used as follows.
+
+```python
+with viewer.sync_slicing():
+    viewer.dims.current_step = (10, 0, 0)
+viewer.screenshot(...)
```

Comment:
Neat! We could even have `screenshot()` default to this behaviour, since typically people don't want a screenshot of a half-loaded canvas.

Reply:
Makes sense to me

Reply:
I was hoping that the implementation of `screenshot` could do that, but after a bit more thought, I don't think it's possible with the proposal here (and maybe more generally with any async slicing).

That's because if `viewer.screenshot` is called from the main thread, it's unclear what it would wait for. The mutation to `current_step` has already occurred on the main thread and triggered an async task to start executing on the slicing thread. `viewer.screenshot` could wait for `LayerSlicer` to be idle (i.e. no pending tasks), but that's not quite what we want - we actually want to wait for the vispy canvas to be redrawn. But the redraw has to happen on the main thread - that's why it's queued up on the main event loop in the current proposal. 

I think we'd need some way of enqueuing the screenshot to occur after the redraw (maybe be connecting `LayerSlice.events.ready` to another callback, which sounds more complicated and fragile than I'd like.

Reply:
I'm not sure I follow... Why isn't connecting to the "ready" event enough? That will happen in the main thread, which means it will be after redraw, right?

Reply:
`LayerSlice.events.ready` is not guaranteed to be executed/emitted in the main thread, because [`add_done_callback` does not guarantee in which thread each callback is executed](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future.add_done_callback). Based on logging, it typically runs in the non-main/slicing thread, and only runs in the main thread when the async task is already done before the callback is added.

That's why we need to add the Qt specific `@ensure_main_thread` on `QtViewer.on_slice_ready`. In the case that a screenshot is requested while the slicer has a pending slice task, I guess we could add a pending screenshot that would actually be executed at the end of `QtViewer.on_slice_ready`. Though, this behavior might be surprising for some - maybe they'd just expect a screenshot to capture what the canvas looks like at that specific time?

Reply:
hm. Maybe processEvents https://doc.qt.io/qt-5/qcoreapplication.html#processEvents may allow doing spin lock until the rendering is done?

Reply:
I assumed that `ready` referred to a `ready` event fired by the *Viewer* once it received the slice and has sent it to `vispy` (which afaik is all happening on the main thread by your diagrams). We definitely should't attach to a `ready` event sent by the `slice` object, since "being done with the slicing computations" does not equate to "being done updating the viewer to the new requested state".

Am I misunderstanding the issue here? I feel like we don't need any fancy machinery; simply have a `ready` event be fired by the viewer model once all the requested slices have been sent to vispy and there is no more pending thread slicing other stuff.

Reply:
> Am I misunderstanding the issue here? I feel like we don't need any fancy machinery; simply have a `ready` event be fired by the viewer model once all the requested slices have been sent to vispy and there is no more pending thread slicing other stuff.

I'll try to clarify here, but it also might be easier to do on a call.

The `ViewerModel` doesn't know anything about Qt and vispy, which seems like a deliberate napari-wide design choice and something we should not change. All it can do is emit an event (in general on the slicing thread) with the slice response data to anything that cares (i.e. anything that has connected to that event).

Instead, `QtViewer.on_slice_ready` is connected to `LayerSlicer.events.ready`, which ensures that all updates (including vispy/qwidget updates) are done on the main thread.

We can add a different event (let's call it 'rendered') to be emitted at the end of `QtViewer.on_slice_ready` which will be emitted on the main thread (since that method is decorated with `@ensure_main_thread`). *But what should connect to this event? And when should it connect?*

We obviously don't want `screenshot` to always be connected to that event. So I assume the plan is to connect to it in `QtViewer.screenshot` with a callback that will actually execute the screenshot, but only when there is a pending slice task - this is the slight tricky part that will require a lock (in `LayerSlicer`) since we need to make sure that the slicing thread does not finish as we are inspecting this state. The screenshot callback should also disconnect itself from the event (i.e. it should act like a single shot event). That effectively what I was suggested above with the pending screenshot, but using an event callback.

Reply:
I can also try to modify one of the existing diagrams to include `screenshot` and clarify the potential timing/order of things happening here, if that would help.

Reply:
I agree that relying on events will be tricky to make the screenshotting robust. 

Perhaps I'm missing something as I'm coming into this conversation late, but would it be possible for the `screenshot()` function to check the state of LayerSlicer  and the visuals to see if there are outstanding requests and if so wait and try again (e.g., use timer thread?)? If the polling time is reasonably short, I think it will be too bad of a user experience if they are unlucky  and request a screenshot just as the last outstanding request is coming in (i.e., they have to wait one polling cycle, which I imaging <= 100 ms). Of course, we would have to add a state variable to mark the visual as stale when a slice request is made for the layer and is set to valid when the new data are received. Both `LayerSlicer` and the current `screenshot` function already requires access to `QtViewer`, so I think this should be possible - not sure it's the best solution though.

Reply:
> The ViewerModel doesn't know anything about Qt and vispy, which seems like a deliberate napari-wide design choice and something we should not change. All it can do is emit an event (in general on the slicing thread) with the slice response data to anything that cares (i.e. anything that has connected to that event).

Right, of course, sorry. I see the issue now :) 

> We obviously don't want screenshot to always be connected to that event.

Rather than an event (or, on top of it), why not have a `rendered` flag that is always `False` during slicing? This way it's fully copntained in the qt side, it can be checked at any time when `screenshot` is fired, and if `False` it can hang until the event is fired.

Maybe I'm saying the same thing as you, in which case ignore my ramblings, I'm still a bit confused about the order of things here :P

PS: if this is an issue, I think it's a pretty good assumption that if someone wants to use the *napari* machinery to screenshot, they mean to screenshot a fully rendered state, and not whatever half-state happens to be rendered in that moment. If one wants the latter (say, for debugging async behaviour), they can always use external tools or a lower level function.

Reply:
(all this is obviously non-blocking for the NAP approval, just to be clear)

Reply:
> In the case that a screenshot is requested while the slicer has a pending slice task, I guess we could add a pending screenshot that would actually be executed at the end of QtViewer.on_slice_ready. Though, this behavior might be surprising for some - maybe they'd just expect a screenshot to capture what the canvas looks like at that specific time?

I agree that some may be expecting to capture the viewer as it is in that moment. However, for people who are doing headless screenshoting (e.g., to make paper figures), it will be important to have some sort of "wait until everything is rendered option" for `viewer.screenshot()`. Perhaps there could be a keyword argument that defaults to waiting for the full scene to be loaded?

In any case, I think this is interesting discussion, but I don't think it's a blocker for accepting this NAP.

Reply:
I added a summary of this conversation to the discussion section, with a link back to this.